regions_www/m/pct_shiny_funs.R

```diff
@@ -0,0 +1,758 @@
+}
+
+
+## Create percentages such that rounded to 1DP if >0% and <0.5%.  [NB can be NaN if denominator zero]
+round_percent <- function(data, expression){
+ value <- round(100 * expression)
+  for(i in 1:length(data)) {
+    if(!is.na(expression[i]) && !is.nan(expression[i]) && expression[i]>0 && expression[i]<0.005) {
+      value[i] <- round(100 * expression[i], 1)
+    }
+    if (is.nan(value[i]) || is.na(value[i])) { value[i] <- "-" }
+  }
+ value
+}
+
+
+## Define and apply the colours of lines
+line_and_colour_df <- data.frame(
+  line_type = c("centroids", "straight_lines", "routes_fast", "routes_quieter", "route_network"),
+  line_colour = c("maroon", "maroon", "purple", "darkgreen", "blue")
+)
+
+get_line_colour <- function(line_type){
+  line_and_colour_df$line_colour[line_and_colour_df$line_type == line_type]
+}
+
+
+## Create customised colour scales
+get_colour_palette <- function(colourscale, bins = 10){
+  # Manually modify to be 'standard 10 plus one extra' for 11 levels
+  if (colourscale == "RdYlBu" && bins == 11) {
+    local_palette <- RColorBrewer::brewer.pal(n = 10, name = colourscale)
+    extra_colour <- "#2d004b"
+    local_palette <- append(local_palette, extra_colour)
+  } else {
+    local_palette <- RColorBrewer::brewer.pal(n = bins, name = colourscale)
+  }
+  # Replace #e0f3f8 with #c6dbef for colourbrewer "RdYlBu"
+  if (colourscale == "RdYlBu") {
+    local_palette <- gsub(pattern = "#E0F3F8", replacement = "#C6DBEF", x = local_palette)
+  }
+  local_palette
+}
+
+
+## Generate a series of colours based on the input values + breaks
+get_colour_ramp <- function(colourscale, bins = 10, values, breaks) {
+  get_colour_palette(colourscale, bins)[cut(x = values, breaks = breaks)]
+}
+
+
+#################
+# LABELLING FUNCTIONS FOR MAP INTERFACE
+#################
+
+# Get scenario name
+scenario_name_df <- data.frame(
+  sc_s_name = c("olc", "govtarget","gendereq","dutch", "ebike"),
+  sc_f_name_commute = c("Census 2011 Cycling", "Government Target", "Gender equality", "Go Dutch", "Ebikes"),
+  sc_f_name_school = c("School Census 2011", "Government Target", "Gender equality", "Go Dutch", "Ebikes"),
+  sc_f_name_alltrips = c("Current travel patterns", "Government Target", "Gender equality", "Go Dutch", "Ebikes")
+)
+
+get_scenario_name <- function(scenario_name, purpose){
+  if (purpose=="commute") {scenario_name_df$sc_f_name_commute[scenario_name_df$sc_s_name == scenario_name]}
+  else if (purpose=="school") {scenario_name_df$sc_f_name_school[scenario_name_df$sc_s_name == scenario_name]}
+  else if (purpose=="alltrips") {scenario_name_df$sc_f_name_alltrips[scenario_name_df$sc_s_name == scenario_name]}
+}
+
+# Get names in pop-ups (variable by purpose)
+text_zone_header <- function(purpose) {
+  if(purpose=="commute") {"All residents living in zone"}
+  else if(purpose=="school") {"All children living in zone"}
+  else if(purpose=="alltrips") {"All trips starting in zone"}
```

Comment:
Nice, a `switch` might be cleaner:

```R
switch(purpose,
  "commute" = "All residents living in zone",
  "school" = "All children living in zone",
  "alltrips" = "All trips starting in zone"
)
```

Reply:
Agreed. Clear improvement, good refactor but it's not causing the website to crash so not a priority atm.

Reply:
Not a priority but where else do we comment on things?

I'm going though the PR line by line and seeing what changes there are.

Reply:
yyes, agreed, that looks  much nicer!